sdk/cosmos/azure_data_cosmos/src/location_cache.rs

```diff
@@ -0,0 +1,570 @@
+                (
+                    "Location 2".to_string(),
+                    "https://location2.documents.azure.com".to_string()
+                ),
+                (
+                    "Location 3".to_string(),
+                    "https://location3.documents.azure.com".to_string()
+                ),
+                (
+                    "Location 4".to_string(),
+                    "https://location4.documents.azure.com".to_string()
+                )
+            ])
+        );
+
+        assert_eq!(
+            cache.locations_info.write_endpoints,
+            vec![
+                "https://location1.documents.azure.com".to_string(),
+                "https://location2.documents.azure.com".to_string()
+            ]
+        );
+
+        assert_eq!(
+            cache.locations_info.read_endpoints,
+            vec![
+                "https://location1.documents.azure.com".to_string(),
+                "https://location2.documents.azure.com".to_string(),
+            ]
+        );
+    }
+
+    #[test]
+    fn location_cache_update_with_one_preferred_location() {
+        let (default_endpoint, write_locations, read_locations, preferred_locations) =
+            create_test_data();
+
+        let mut cache = LocationCache::new(
+            default_endpoint.to_string(),
+            vec![preferred_locations[0].clone()],
+        );
+
+        cache.update(write_locations, read_locations);
+
+        assert_eq!(
+            cache.default_endpoint,
+            "https://default.documents.azure.com"
+        );
+
+        assert_eq!(
+            cache.locations_info.preferred_locations,
+            vec![preferred_locations[0].clone()]
+        );
+
+        assert_eq!(
+            cache.locations_info.write_endpoints,
+            vec!["https://location1.documents.azure.com".to_string()]
+        );
+
+        assert_eq!(
+            cache.locations_info.read_endpoints,
+            vec!["https://location1.documents.azure.com".to_string()]
+        );
+    }
+
+    #[test]
+    fn mark_read_endpoint_unavailable_test() {
+        // set up test data
+        let (default_endpoint, write_locations, read_locations, preferred_locations) =
+            create_test_data();
+
+        let mut cache = LocationCache::new(default_endpoint, preferred_locations);
+        cache.update(write_locations, read_locations);
+
+        // mark location 1 as unavailable endpoint for read operation
+        let unavailable_endpoint = "https://location1.documents.azure.com";
+        let operation = RequestOperation::Read;
+        cache.mark_endpoint_unavailable(unavailable_endpoint, operation);
+
+        // check that endpoint is last option in read endpoints and it is in the location unavailability info map
+        assert_eq!(
+            cache.locations_info.read_endpoints.last(),
+            Some(&unavailable_endpoint.to_string())
+        );
+
+        assert!(cache.is_endpoint_unavailable(unavailable_endpoint, operation));
```

Comment:
Add an assert on the result of `resolve_service_endpoint` to this test (and the one below) as well. We know that if `read_endpoints` is correct, `resolve_service_endpoint` _should be as well_ but `resolve_service_endpoints` is the actual API that users of this type will see, so we should assert that to make it clear.